### PR TITLE
mdb with tls/ssl encryption and so much more!

### LINKED ISSUE #23: gdbserver is not secure
Due to an underlying security issue with `gdbserver`, it is **not recommended** to run `mdb` on multi-user systems connected to a public network e.g., most HPC systems.

> Warning: gdbserver does not have any built-in security. Do not run gdbserver connected to any public network; a GDB connection to gdbserver provides access to the target system with the same privileges as the user running gdbserver.

(source [gdb manual](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Server.html))

I am working on a fix which involves replacing the `gdbserver` backend with my own `tls`/`ssl`-encrypted server to manage connections and directly use `gdb` as a backend instead.

I am planning to use randomly-generated tokens (similar to `jupyter notebook`) to secure the connection.

- [ ] replace `gdbserver` backend with `tls`/`ssl`-encrypted server + `gdb` backend
- [ ] update documentation and remove security notice

### PR DESCRIPTION
**Pull request by TomMelt**
This is a mega merge :flushed: to address the security issue in #23.

I will try not to do this ever again. So many changes possibly too many to mention.

Completely stripped out the backend (`gdbserver`) and fundamental way this app works.

We (@fjebaker and I) added asyncio client server backend which is documented [here](https://mdb.readthedocs.io/en/latest/client-server.html).

I have updated the [quickstart guide](https://mdb.readthedocs.io/en/latest/quick-start.html) to reflect most of these changes.

Still to do: 
- [ ] fix the CI test which doesn't run currently (it runs successfully locally but not on GitHub) 
- [ ] create docker image to run code (to remove issues with `gdb`/`lldb` versions)
- [ ] test on multinode run
 

### ToDo
- [x] update docs
- [x] add llvm test
- [x] check coverage

closes #23

### CODE REVIEW

File: mdb/exchange_server.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            msg = await conn.recv_message()
+        except Exception as e:
+            print(e)
+        logger.info(
+            "exchange server received {%s} from {%s}.",
+            msg.msg_type,
+            msg.data["from"],
+        )
+
+        # checkhov's loop
```
Comment (fjebaker) [inline mdb/exchange_server.py:59]:
```suggestion
        # chekhov's loop
```
I am shocked I misspelled his name

File: mdb/mdb_launch.py [hunk 2]
```diff
@@ -56,10 +66,23 @@
...
+    type=click.File("r"),
+    required=True,
+    help="Target binary to debug.",
+)
+@click.option(
+    "--mdb-home",
+    default="~/.mdb",
     show_default=True,
-    help="filename for the mpirun configuration.",
+    help="filename for the mpirun appfile.",
```
Comment (fjebaker) [inline mdb/mdb_launch.py:85]: Is this correct for the `--mdb-home` options? I thought that was where e.g. the certs and keys are stored?

File: mdb/mdb_launch.py [hunk 3]
```diff
@@ -90,27 +122,69 @@ def launch(
...
+
+    MDB_HOME = expanduser(mdb_home)
+    MDB_CERT_PATH = join(MDB_HOME, "cert.pem")
+    MDB_KEY_PATH = join(MDB_HOME, "key.rsa")
+
+    if not exists(MDB_HOME):
+        mkdir(MDB_HOME)
+
+    if not (exists(MDB_CERT_PATH) or exists(MDB_KEY_PATH)):
+        subj = "/C=XX/ST=mdb/L=mdb/O=mdb/OU=mdb/CN=localhost"
```
Comment (fjebaker) [inline mdb/mdb_launch.py:141]:
```suggestion
        subj = f"/C=XX/ST=mdb/L=mdb/O=mdb/OU=mdb/CN={hostname}"
```
Should be consistent with the hostname given on the CLI.

File: mdb/messages.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    @staticmethod
+    def from_json(text: bytes) -> "Message":
+        msg = json.loads(text.decode()[: -len(END_BYTES)])
+        return Message(msg["msg_type"], msg["data"])
+
+    @staticmethod
+    def debug_conn_request() -> "Message":
+        return Message(
+            "debug conn request",
```
Comment (fjebaker) [inline mdb/messages.py:28]: This is just my personal preference, but I would recommend avoiding spaces in the message types, and use either hierarchical `"debug.conn.request"`, else snake case `"debug_conn_request"` as it makes finding them and filtering in the logs a bit easier.

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -12,17 +12,22 @@ jobs:
...
         sudo apt update
-        sudo apt install -yq gdb gdbserver libopenmpi-dev
+        sudo apt install -y gdb libopenmpi-dev libssl-dev
     - name: Build example mpi program
       run: |
         make -C examples/
     - name: Install mdb and additional python packages
       run: |
+        mkdir -p ~/.mdb
+        openssl req -x509 -newkey rsa:4096 -sha256 -days 365 -keyout ~/.mdb/key.rsa -out ~/.mdb/cert.pem -nodes -subj "/C=XX/ST=mdb/L=mdb/O=mdb/OU=mdb/CN=localhost"
```
Comment (fjebaker) [inline .github/workflows/tests.yml:29]: Shouldn't the exchange server do this when it starts up?

**Review by fjebaker — COMMENTED** [review]
Very cool stuff! The regression suite is nice too! I just noticed a few bits and bobs when looking over the code again

File: mdb/async_connection.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        self.writer = writer
+        self.type = ""
+
+    async def recv_message(self) -> "Message":
+        msg = Message.from_json(await self.reader.readuntil(separator=END_BYTES))
+        logger.debug("msg received [%s]", msg.msg_type)
+        return msg
+
+    async def send_message(self, msg: Message) -> None:
+        self.writer.write(msg.to_json())
```
Comment (fjebaker) [inline mdb/async_connection.py:28]:
```suggestion
        try:
             self.writer.write(msg.to_json())
        except Exception as e:
            logger.exception("async send error")
            raise e
```

File: mdb/async_connection.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        self,
+        reader: asyncio.StreamReader,
+        writer: asyncio.StreamWriter,
+    ):
+        self.reader = reader
+        self.writer = writer
+        self.type = ""
+
+    async def recv_message(self) -> "Message":
+        msg = Message.from_json(await self.reader.readuntil(separator=END_BYTES))
```
Comment (fjebaker) [inline mdb/async_connection.py:23]:
```suggestion
        try:
             raw_msg = await self.reader.readuntil(separator=END_BYTES)
        except Exception as e:
            logger.exception("async read error")
            raise e
        msg = Message.from_json(raw_msg)
```

**Review by fjebaker — COMMENTED** [review]
This might help with debugging the CI errors?

File: mdb/async_client.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    def _init_tls(self) -> None:
+        context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+        context.load_cert_chain(
+            ssl_cert_path(),
+            ssl_key_path(),
+        )
+        context.load_verify_locations(ssl_cert_path())
+
+        # insecure debug mode
+        if os.environ.get("MDB_DISABLE_HOSTNAME_VERIFY", None):
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply mdb/async_client.py:33]: they are now... plus I added a bonus feature which allows you to switch and CLI option for an environment variable (see [docs/source/environment-vars.rst](https://github.com/TomMelt/mdb/blob/e422afc8d42d5539e19059380285bb15c59d08e0/docs/source/environment-vars.rst))

File: mdb/mdb_launch.py (hunk 2 above)
Comment (TomMelt) [reply mdb/mdb_launch.py]: :flushed: nope. I have fixed this now

File: mdb/messages.py (hunk 4 above)
Comment (TomMelt) [reply mdb/messages.py]: good idea :bulb: 

I decided on snake case ... because python :snake:

File: .github/workflows/tests.yml (hunk 5 above)
Comment (TomMelt) [reply .github/workflows/tests.yml:29]: it should... this is from my debugging. I will leave this unresolved until I remove this garbage

File: mdb/async_client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fjebaker) [reply mdb/async_client.py:33]: 🚀🚀🚀

File: .github/workflows/tests.yml (hunk 5 above)
Comment (TomMelt) [reply .github/workflows/tests.yml]: I have removed it :+1:

File: mdb/debug_client.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+            if not success:
+                # nothing needed cancelling, so no reply needed
+                logger.debug("No task to interrupt")
+
+            # send intterupt to the process
+            self.dbg_proc.sendintr()
+            await self.dbg_proc.expect(self.backend.prompt_string, async_=True)
+            # report on how that all went
+            output = f"\r\nInterrupted: {success}\r\n"
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply mdb/debug_client.py:63]: raised issue #32 to address this... but I would like to get rid of this mega PR before it gets any bigger :grimacing:

**Pull request merged**